src/cpu/aarch64/acl_gemm_convolution.hpp

```diff
@@ -0,0 +1,194 @@
+            acp.with_bias ? &acp.bia_info : nullptr,
+            &acp.dst_info,
+            acp.padstride_info,
+            acp.weights_info,
+            acp.dilation_info,
+            acp.act_info);
+        // clang-format on
+        if (acl_st.error_code() != arm_compute::ErrorCode::OK) {
+            return status::unimplemented;
+        }
+
+        // Init Compute Library tensors based on info from descriptor
+        acl_obj_->src_tensor.allocator()->init(acp.src_info);
+        acl_obj_->wei_tensor.allocator()->init(acp.wei_info);
+        acl_obj_->dst_tensor.allocator()->init(acp.dst_info);
+        acl_obj_->bia_tensor.allocator()->init(acp.bia_info);
+        if (acp.with_bias) {
+            acl_obj_->bia_memory = std::calloc(acp.bia_info.tensor_shape()[0],
+                    sizeof(acp.bia_info.data_type()));
+        }
+
+        // clang-format off
+        acl_obj_->gemm_conv.configure(
+            &acl_obj_->src_tensor,
+            &acl_obj_->wei_tensor,
+            acp.with_bias ? &acl_obj_->bia_tensor : nullptr,
+            &acl_obj_->dst_tensor,
+            acp.padstride_info,
+            acp.weights_info,
+            acp.dilation_info,
+            acp.act_info);
+        // clang-format on
+
+        return status::success;
+    }
+
+    acl_obj_t &get_acl_obj() const { return *acl_obj_; }
+
+    DNNL_DISALLOW_COPY_AND_ASSIGN(acl_resource_t);
+
+private:
+    std::unique_ptr<acl_obj_t> acl_obj_;
+
+}; // acl_resource_t
+
+struct acl_gemm_convolution_fwd_t : public primitive_t {
+    struct pd_t : public cpu_convolution_fwd_pd_t {
+        pd_t(const convolution_desc_t *adesc, const primitive_attr_t *attr,
+                const typename pd_t::base_class *hint_fwd_pd)
+            : cpu_convolution_fwd_pd_t(adesc, attr, hint_fwd_pd), acp_() {}
+
+        DECLARE_COMMON_PD_T(GEMM_IMPL_STR, acl_gemm_convolution_fwd_t,
+                USE_GLOBAL_SCRATCHPAD);
+
+        status_t init(engine_t *engine) {
+            bool ok = true && is_fwd()
+                    && set_default_alg_kind(alg_kind::convolution_direct)
+                    && expect_data_types(data_type::f32, data_type::f32,
+                            data_type::f32, data_type::f32, data_type::f32)
+                    && !has_zero_dim_memory()
+                    && attr()->has_default_values(
+                            primitive_attr_t::skip_mask_t::post_ops)
+                    && post_ops_ok();
+            if (!ok) return status::unimplemented;
+
+            auto conf_status = acl_gemm_convolution_utils::init_conf(acp_,
+                    src_md_, weights_md_, dst_md_, bias_md_, *desc(), *attr());
+            if (conf_status != status::success) return status::unimplemented;
+
+            // Number of threads in Compute Library is set by OMP_NUM_THREADS
+            // dnnl_get_max_threads() == OMP_NUM_THREADS
+            arm_compute::Scheduler::get().set_num_threads(
+                    dnnl_get_max_threads());
```

Comment:
@emfomenk , good morning. Many thanks for the detailed explanation.
> Could you please elaborate on what actually happens here? 

The idea is to restrict from above the number of pthreads in use by Arm Compute Library functions. Unfortunately, it can not be done with an environmental variable like in the case of OMP scheduler.

> What would happen if we remove this snippet completely?

Then ACL will create by default number_of_cores pthreads, and will occupy all the cores in the calls to `arm_gemm_convolution_fwd_t`.

> Looking at this chunk it seems that arm_compute::Scheduler::get() is a global object. Could it be a problem for the scenario above?

Yes, [arm_compute::Scheduler](https://github.com/ARM-software/ComputeLibrary/blob/master/arm_compute/runtime/Scheduler.h) is global, and is created for the first time when a function, referring to it, is invoked. There will be no problems in the first example, `p.execute(...);` will employ 10 POSIX threads. The only potential problem is that the last call to `arm_compute::Scheduler::get().set_num_threads(...)` will define the maximal number of threads to use in _all_ future ACL-related primitives executions since `Scheduler` is global. It was assumed here, that `dnnl_get_max_threads()` always returns the same number. At the moment there seems to be no way to tune different primitives to use different number of threads in ACL, only a global limit is supported.

Reply:
@alenik01, thank you for the clarification!
Alright, that's a limitation for now. I just want to make sure we are aware of it. I guess in most real cases `dnnl_get_max_threads()` indeed returns the same number, so we should be fine.

Resolving the discussion.